ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java

```diff
@@ -316,6 +314,30 @@ private boolean removeFiles(String location, ValidWriteIdList writeIdList, Compa
       }
       fs.delete(dead, true);
     }
-    return true;
+    // Check if there will be more obsolete directories to clean when possible. We will only mark cleaned when this
+    // number reaches 0.
+    return getNumEventuallyObsoleteDirs(location, dirSnapshots) == 0;
+  }
+
+  /**
+   * Get the number of base/delta directories the Cleaner should remove eventually. If we check this after cleaning
+   * we can see if the Cleaner has further work to do in this table/partition directory that it hasn't been able to
+   * finish, e.g. because of an open transaction at the time of compaction.
+   * We do this by assuming that there are no open transactions anywhere and then calling getAcidState. If there are
+   * obsolete directories, then the Cleaner has more work to do.
+   * @param location location of table
+   * @return number of dirs left for the cleaner to clean – eventually
+   * @throws IOException
+   */
+  private int getNumEventuallyObsoleteDirs(String location, Map<Path, AcidUtils.HdfsDirSnapshot> dirSnapshots)
+      throws IOException {
+    ValidTxnList validTxnList = new ValidReadTxnList();
```

Comment:
Will we consider all writes as valid here?

Reply:
Compactor txnid is stored since HIVE-24291. This patch is for versions without that change, or any recent changes to the hms schema.

Reply:
Could we limit by CQ_NEXT_TXN_ID ?  
Do we need this change in upstream master if HIVE-24291 is already present ?

Reply:
No we don't need this change if HIVE-24291 is present. But if HIVE-24291 is present it shouldn't hurt.
CQ_NEXT_TXN_ID is stored since HIVE-23107, which also has hms schema changes.

Reply:
> But if HIVE-24291 is present it shouldn't hurt.
This isn't necessarily true, since as @pvargacl noted there could be aborts since the compaction happened.
The question here is whether this change or HIVE-24314 is worse for users if HIVE-23107 (remove MIN_HISTORY_LEVEL) etc. are not present on their version.